src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+                        break;
+                    case SyntaxKind.ReturnStatement:
+                        if (permittedJumps & PermittedJumps.Return) {
+                            facts |= RangeFacts.HasReturn;
+                        }
+                        else {
+                            (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalReturnStatement));
+                        }
+                        break;
+                    default:
+                        forEachChild(n, visit);
+                        break;
+                }
+
+                permittedJumps = savedPermittedJumps;
+            }
+        }
+    }
+
+    /**
+     * Computes possible places we could extract the function into. For example,
+     * you may be able to extract into a class method *or* local closure *or* namespace function,
+     * depending on what's in the extracted body.
+     */
+    export function collectEnclosingScopes(range: TargetRange): Scope[] {
+        let current: Node = isArray(range.range) ? firstOrUndefined(range.range) : range.range;
+        if (range.facts & RangeFacts.UsesThis) {
+            // if range uses this as keyword or as type inside the class then it can only be extracted to a method of the containing class
+            const containingClass = getContainingClass(current);
+            if (containingClass) {
+                return [containingClass];
+            }
+        }
+
+        const scopes: Scope[] = [];
+        while (current) {
+            // We want to find the nearest parent where we can place an "equivalent" sibling to the node we're extracting out of.
+            // Walk up to the closest parent of a place where we can logically put a sibling:
+            //  * Function declaration
+            //  * Class declaration or expression
+            //  * Module or source file
+            // Note that we don't use isFunctionLike because we don't want to put the extracted closure *inside* a method
+            if ((current.kind === SyntaxKind.FunctionDeclaration) || isSourceFile(current) || isModuleBlock(current) || isClassLike(current)) {
+                scopes.push(current as FunctionLikeDeclaration);
+            }
+            current = current.parent;
+        }
+        return scopes;
+    }
+
+    /**
+     * Given a piece of text to extract ('targetRange'), computes a list of possible extractions.
+     * Each returned ExtractResultForScope corresponds to a possible target scope and is either a set of changes
+     * or an error explaining why we can't extract into that scope.
+     */
+    export function extractRange(targetRange: TargetRange, sourceFile: SourceFile, context: RefactorContext): ReadonlyArray<ExtractResultForScope> {
+        if (targetRange === undefined) {
+            return [];
+        }
+
+        const scopes = collectEnclosingScopes(targetRange);
+        if (scopes.length === 0) {

```

Comment:
Rather than allocate an empty array in `collectEnclosingScopes`, compare its length, and then allocate _another_ empty array here, we should probably return `undefined` from `collectEnclosingScopes`, just test `if (!scopes)` and return `undefined`.